File: browser_prototype/src/entity.ts
class InvalidParameter extends Error { }

class Coordinate implements IHashable, ISerializable, ICopyable<Coordinate>
{
    constructor(public x: number, public y: number)
    {
        if (!Coordinate.is_valid(x, y))
        {
            throw new InvalidParameter("Coordinate");
        }
    }

    equals(other: Coordinate): boolean
    {
        return this.x == other.x && this.y == other.y;
    }

    add(dx: number, dy: number): Coordinate | null
    {
        if (!Coordinate.is_valid(this.x + dx, this.y + dy))
        {
            return null;
        }
        return new Coordinate(this.x + dx, this.y + dy);
    }

    copy(): Coordinate
    {
        return new Coordinate(this.x, this.y);
    }

    hash(): string
    {
        return `Coordinate(${ this.x },${ this.y })`;
    }

    serialize(): string
    {
        return `${ this.x }${ this.y }`;
    }

    static deserialize(payload: string): Coordinate
    {
        return new Coordinate(parseInt(payload[0]), parseInt(payload[1]));
    }

    static is_valid(x: number, y: number): boolean
    {
        return 0 <= x && x < g.board_size_x && 0 <= y && y < g.board_size_y;
    }
}

class Skill implements IHashable
{
    constructor(public x: number, public y: number)
    {
        if (!Skill.is_valid(x, y))
        {
            throw new InvalidParameter("Skill");
        }
    }

    hash(): string 
    {
        return `Skill(${ this.x },${ this.y })`;
    }

    equals(other: Skill): boolean
    {
        return this.x == other.x && this.y == other.y;
    }

    is_leap(): boolean
    {
        return Math.abs(this.x) > 1 || Math.abs(this.y) > 1;
    }

    static is_valid(x: number, y: number)
    {
        return -g.skill_range <= x && x <= g.skill_range &&
            -g.skill_range <= y && y <= g.skill_range;
    }
}

class SkillSet implements ISerializable, ICopyable<SkillSet>
{
    private map: boolean[];

    constructor(skills: Skill[] = [])
    {
        this.map = [];
        for (let i = 0; i < g.skillset_size * g.skillset_size; i++)
        {
            this.map[i] = false;
        }

        for (let skill of skills)
        {
            this.add(skill);
        }
    }

    add(skill: Skill)
    {
        this.map[(skill.x + g.skill_range) * g.skillset_size + skill.y + g.skill_range] = true;
    }

    has(skill: Skill): boolean
    {
        return this.map[(skill.x + g.skill_range) * g.skillset_size + skill.y + g.skill_range];
    }

    serialize(): string
    {
        let map: number[] = [];

        for (let i = 0; i < g.skillset_size * g.skillset_size; i++)
        {
            map[i] = this.map[i] ? 1 : 0;
        }

        return JSON.stringify(map);
    }

    static deserialize(payload: string): SkillSet
    {
        let map = JSON.parse(payload);
        let sks = new SkillSet();
        for (let i = 0; i < g.skillset_size * g.skillset_size; i++)
        {
            sks.map[i] = map[i] == 1;
        }
        return sks;
    }

    static from_literal(matrix: string): SkillSet
    {
        let rows = matrix.split('\n');
        rows = rows.map((row: string) => { return row.trim(); });
        let skills: Skill[] = [];
        for (let dy = -g.skill_range; dy <= g.skill_range; dy++)
        {
            for (let dx = -g.skill_range; dx <= g.skill_range; dx++)
            {
                if (rows[dy + g.skill_range][dx + g.skill_range] == 'x')
                {
                    skills.push(new Skill(dx, dy));
                }
            }
        }

        return new SkillSet(skills);
    }

    as_list(): Skill[]
    {
        let skills: Skill[] = [];
        for (let dy = -g.skill_range; dy <= g.skill_range; dy++)
        {
            for (let dx = -g.skill_range; dx <= g.skill_range; dx++)
            {
                if (this.map[(dx + g.skill_range) * g.skillset_size + dy + g.skill_range])
                {
                    skills.push(new Skill(dx, dy));
                }
            }
        }

        return skills;
    }

    copy(): SkillSet
    {
        return new SkillSet().union(this);
    }

    static op_union = (a: boolean, b: boolean) => a || b;
    static op_subtract = (a: boolean, b: boolean) => a && !b;

    apply(operator: (a: boolean, b: boolean) => boolean, other: SkillSet): SkillSet
    {
        let result = new SkillSet();

        for (let i = 0; i < g.skillset_size * g.skillset_size; i++)
        {
            result.map[i] = operator(this.map[i], other.map[i]);
        }

        return result;
    }

    union(other: SkillSet): SkillSet
    {
        return this.apply(SkillSet.op_union, other);
    }

    subtract(other: SkillSet): SkillSet
    {
        return this.apply(SkillSet.op_subtract, other);
    }

    equals(other: SkillSet): boolean
    {
        for (let i = 0; i < g.skillset_size * g.skillset_size; i++)
        {
            if (this.map[i] != other.map[i])
            {
                return false;
            }
        }
        return true;
    }

    flip(): SkillSet
    {
        let flipped = this.copy();

        for (let i = 0; i < g.skillset_size; i++)
        {
            for (let j = 0; j < g.skill_range; j++)
            {
                [
                    flipped.map[i * g.skillset_size + j],
                    flipped.map[i * g.skillset_size + g.skillset_size - 1 - j]
                ] = [
                        flipped.map[i * g.skillset_size + g.skillset_size - 1 - j],
                        flipped.map[i * g.skillset_size + j]
                    ];
            }
        }
        return flipped;
    }
}

enum Player
{
    P1 = 1,
    P2 = 2
}

module Player
{
    export function* both()
    {
        yield Player.P1;
        yield Player.P2;
    }

    export function* values<T>(players: Players<T>)
    {
        yield players[Player.P1];
        yield players[Player.P2];
    }
}

type Players<T> =
    {
        [Player.P1]: T,
        [Player.P2]: T,
    };

module Players
{
    export function create<T>(ctor: (p: Player) => T): Players<T>
    {
        return {
            [Player.P1]: ctor(Player.P1),
            [Player.P2]: ctor(Player.P2)
        };
    }
}

function opponent(player: Player)
{
    return player == Player.P1 ? Player.P2 : Player.P1;
}

function deserialize_player(payload: string): Player
{
    return <Player> Player[<keyof typeof Player>('P' + payload)];
}

function serialize_player(player: Player)
{
    return JSON.stringify(player);
}

class Move implements ISerializable, ICopyable<Move>, IHashable
{
    constructor(public from: Coordinate, public to: Coordinate)
    {
    }

    equals(other: Move): boolean
    {
        return this.from.equals(other.from) && this.to.equals(other.to);
    }

    which_skill(): Skill | null
    {
        let dx = this.to.x - this.from.x;
        let dy = this.to.y - this.from.y;
        if (Skill.is_valid(dx, dy))
        {
            return new Skill(dx, dy);
        }
        else
        {
            return null;
        }
    }

    hash(): string
    {
        return `Move(${ this.from.hash() },${ this.to.hash() })`;
    }

    serialize(): string
    {
        return this.from.serialize() + this.to.serialize();
    }

    static deserialize(payload: string)
    {
        return new Move(Coordinate.deserialize(payload.slice(0, 2)), Coordinate.deserialize(payload.slice(2, 4)));
    }

    copy(): Move
    {
        return new Move(this.from.copy(), this.to.copy());
    }
}

class PlayerMove implements ISerializable
{
    constructor(public player: Player, public moves: Move[] = [])
    {
    }

    append(move: Move): void
    {
        this.moves.push(move);
    }

    pop(): void
    {
        this.moves.pop();
    }

    extract(filter: (m: Move) => m is Move): Move[]
    {
        return extract(this.moves, filter);
    }

    serialize(): string
    {
        let s = this.moves.map(m => m.serialize());
        s.unshift(serialize_player(this.player));
        return JSON.stringify(s);
    }

    static deserialize(payload: string): PlayerMove
    {
        let data = JSON.parse(payload);
        let player = deserialize_player(data.shift());
        let moves = data.map((m: string) => Move.deserialize(m));
        return new PlayerMove(player, moves);
    }
}

class Action implements ICopyable<Action>
{
    constructor(public move: Move, public type: ActionType, public unit_type: UnitConstructor)
    {
    }

    copy(): Action
    {
        return new Action(this.move.copy(), this.type, this.unit_type);
    }

    cost(): number
    {
        function move_cost(unit_type: UnitConstructor, move: Move)
        {
            if (unit_type == Archer)
            {
                return 2;
            }

            if (move.which_skill()!.is_leap())
            {
                return 3;
            }
            else
            {
                return 2;
            }
        }

        switch (this.type)
        {
            case ActionType.Defend:
                return 1;
            case ActionType.Move:
                return move_cost(this.unit_type, this.move);
            case ActionType.Upgrade:
                return 5;
            case ActionType.Attack:
                return move_cost(this.unit_type, this.move) + 1;
        }
    }

    static deserialize(payload: string): Action
    {
        let action_type: string;
        let unit_type: string;
        let move_literal: string;
        [action_type, unit_type, move_literal] = JSON.parse(payload);
        return new Action(Move.deserialize(move_literal), parseInt(action_type), g.unit_type_by_name.get(unit_type)!);
    }
}

enum ActionType
{
    Upgrade = 1,
    Defend = 2,
    Move = 3,
    Attack = 4
}

class PlayerAction
{
    constructor(public player: Player, public actions: Action[] = [])
    {
    }

    cost(): number
    {
        return this.actions.map((a) =>
        {
            return a.cost();
        }).reduce((a, b) => a + b, 0);
    }

    static deserialize(payload: string): PlayerAction
    {
        let s = JSON.parse(payload);
        let player = deserialize_player(s.shift());
        let actions = [];
        for (let action_literal of s)
        {
            actions.push(Action.deserialize(action_literal));
        }

        return new PlayerAction(player, actions);
    }

    extract(filter: (a: Action) => a is Action): Action[]
    {
        return extract(this.actions, filter);
    }
}

abstract class Unit implements ISerializable, ICopyable<Unit>
{
    readonly perfect: SkillSet;
    current: SkillSet;
    readonly promotion_options: AdvancedUnitConstructor[] = [];
    readonly level: number = 0;

    constructor(public owner: Player, current: SkillSet | null = null)
    {
        this.perfect = g.perfect_skills[this.type().id];
        this.current = current == null ? new SkillSet() : current;
    }

    serialize(): string
    {
        return JSON.stringify([this.type().name, this.owner, this.current.serialize()]);
    }

    copy(): Unit
    {
        let ctor = <UnitConstructor> this.constructor;
        let u = new ctor(this.owner, this.current.copy());
        return u;
    }

    endow_inborn(): void
    {
        let inborn = g.inborn_skills[this.type().id];
        if (!inborn)
        {
            return;
        }
        this.current = this.owner == Player.P1 ? inborn.copy() : inborn.flip();
    }

    type(): UnitConstructor
    {
        return <UnitConstructor> this.constructor;
    }

    endow(skill: Skill): boolean
    {
        if (this.perfect.has(skill))
        {
            this.current.add(skill);
            return true;
        }
        return false;
    }

    capable(skill: Skill): boolean
    {
        return this.current.has(skill);
    }

    potential(): SkillSet
    {
        return this.perfect.subtract(this.current);
    }

    is_promotion_ready(): boolean
    {
        return false;
    }

    is_perfect(): boolean
    {
        return this.current.equals(this.perfect);
    }

    is_advanced(): boolean
    {
        return false;
    }

    duel(other: Unit): Unit | null
    {
        if (this.level == other.level)
        {
            return null;
        }
        else if (this.level > other.level)
        {
            return this;
        }
        else
        {
            return other;
        }
    }

    promote(skill: Skill): Unit | null
    {
        if (!this.is_promotion_ready())
        {
            return null;
        }

        let ctor = Unit.which_has_skill(skill, this.promotion_options);
        if (ctor == null)
        {
            return null;
        }

        let promoted = new (<AdvancedUnitConstructor> ctor)(this.owner, this.current);
        promoted.endow(skill);
        return promoted;
    }

    static spawn_perfect(player: Player, ctor: UnitConstructor): Unit
    {
        let unit = new ctor(player, null);
        unit.perfect.as_list().forEach(s => { unit.endow(s); });
        return unit;
    }

    static which_has_skill(skill: Skill, ctors: UnitConstructor[]): UnitConstructor | null
    {
        for (let c of ctors)
        {
            if (g.perfect_skills[c.id].has(skill))
            {
                return c;
            }
        }
        return null;
    }
}

interface UnitConstructor extends IDeserializable<Unit>
{
    new(owner: Player, current: SkillSet | null): Unit;
    deserialize(payload: string): Unit;
    readonly id: number;
    discriminator: string;
}

const UnitConstructor: UnitConstructor = class _ extends Unit
{
    static readonly id = 0;
    static discriminator = '';

    static deserialize(payload: string): Unit
    {
        let display: string, owner: string, current: string;
        [display, owner, current] = <[string, string, string]> JSON.parse(payload);

        let type = g.unit_type_by_name.get(display);
        if (!type)
        {
            throw new Error('Unit.deserialize: no constructor');
        }
        let unit = new type(deserialize_player(owner), SkillSet.deserialize(current));
        return unit;
    }
};

interface BasicUnitConstructor extends UnitConstructor
{
    new(owner: Player, current: SkillSet | null): BasicUnit;
    discriminator: 'BasicUnitConstructor';
}

interface AdvancedUnitConstructor extends UnitConstructor
{
    new(owner: Player, current: SkillSet | null): AdvancedUnit;
    discriminator: 'AdvancedUnitConstructor';
}

function is_basic_unit_ctor(ctor: UnitConstructor): ctor is BasicUnitConstructor 
{
    return 'discriminator' in ctor && ctor['discriminator'] == 'BasicUnitConstructor';
}

function is_advanced_unit_ctor(ctor: UnitConstructor): ctor is AdvancedUnitConstructor 
{
    return 'discriminator' in ctor && ctor['discriminator'] == 'AdvancedUnitConstructor';
}

abstract class BasicUnit extends UnitConstructor
{
    is_promotion_ready(): boolean
    {
        return this.is_perfect();
    }

    potential(): SkillSet
    {
        let potentials = super.potential();
        if (this.is_promotion_ready())
        {
            for (let future_type of this.promotion_options)
            {
                potentials = potentials.union(g.perfect_skills[future_type.id]);
            }
        }

        return potentials;
    }
}

abstract class AdvancedUnit extends UnitConstructor
{
    is_advanced(): boolean
    {
        return true;
    }
}

const AdvancedUnitConstructor: AdvancedUnitConstructor = class _ extends AdvancedUnit
{
    static discriminator: 'AdvancedUnitConstructor' = 'AdvancedUnitConstructor';
};

const BasicUnitConstructor: BasicUnitConstructor = class _ extends BasicUnit
{
    static discriminator: 'BasicUnitConstructor' = 'BasicUnitConstructor';
};

class Rider extends BasicUnitConstructor
{
    static readonly id = 1;
    readonly promotion_options = [Lancer, Knight];
    readonly level = 2;
}

class Soldier extends BasicUnitConstructor
{
    static readonly id = 2;
    readonly promotion_options = [Swordsman, Spearman];
    readonly level = 1;
}

class Archer extends BasicUnitConstructor
{
    static readonly id = 3;
    readonly promotion_options = [Warrior, Spearman];
    readonly level = 1;
}

class Barbarian extends BasicUnitConstructor
{
    static readonly id = 4;
    readonly promotion_options = [Warrior, Swordsman];
    readonly level = 1;
}

class Lancer extends AdvancedUnitConstructor
{
    static readonly id = 5;
    readonly level = 3;
}

class Knight extends AdvancedUnitConstructor
{
    static readonly id = 6;
    readonly level = 3;
}

class Swordsman extends AdvancedUnitConstructor
{
    static readonly id = 7;
    readonly level = 2;
}

class Spearman extends AdvancedUnitConstructor
{
    static readonly id = 8;
    readonly level = 2;
}

class Warrior extends AdvancedUnitConstructor
{
    static readonly id = 9;
    readonly level = 2;
}

class King extends UnitConstructor
{
    static readonly id = 10;
    readonly level = 1;
}

File: browser_prototype/server/entity.py
import skills
from enum import Enum
from const import board_size_x, board_size_y, skillset_size, skillset_range
import json

class Move:
    @classmethod
    def from_literal(self, literal):
        try:
            x1, y1, x2, y2 = literal
            x1, y1, x2, y2 = map(int, [x1, y1, x2, y2])

            position_1 = Position(x1, y1)
            position_2 = Position(x2, y2)
        except:
            raise InvalidParameter("move literal")
        else:
            return Move(position_1, position_2)

    def __init__(self, position_from, position_to):
        self.position_from = position_from
        self.position_to = position_to
    
    def __eq__(self, move):
        return self.position_from == move.position_from and \
            self.position_to == move.position_to

    def is_from_same(self, move):
        return self.position_from == move.position_from

    def is_to_same(self, move):
        return self.position_to == move.position_to

    def get_skill(self):
        return Skill(self.position_from.get_delta(self.position_to))

    def __repr__(self):
        return str(self.position_from) + '->' + str(self.position_to)

    def serialize(self):
        return self.position_from.serialize() + self.position_to.serialize()

class PlayerMove:
    @classmethod
    def from_literal(self, player, literal):
        try:
            return PlayerMove(player, [Move.from_literal(lit) for lit in literal.split()])
        except:
            raise InvalidParameter("player move literal")

    def __init__(self, player, move_list):
        self.player = player
        self.move_list = move_list
    
    def __repr__(self):
        return ','.join([str(m) for m in self.move_list])

class Action:
    def __init__(self, move, type_, unit_type):
        self.move = move
        self.type = type_
        self.unit_type = unit_type

    def __repr__(self):
        return f'{self.move}({ActionType.show(self.type)})'

    def get_cost(self):
        if self.type == ActionType.Move:
            if self.move.get_skill().is_leap():
                return 4
            else:
                return 3
        try:
            return {
                ActionType.Upgrade: 5,
                ActionType.Defend: 2,
                ActionType.Attack: 6,
            }[self.type]
        except KeyError:
            return {
                Barbarian: 8,
                Soldier: 12,
                Archer: 10,
                Rider: 16,
                Wagon: 20
            }[self.unit_type]

    def serialize(self):
        return json.dumps([self.type.value, self.unit_type.__name__, self.move.serialize()])

class PlayerAction:
    def __init__(self, player, action_list):
        self.player = player
        self.action_list = action_list
    
    def extract_actions(self, filter):
        extracted = [a for a in self.action_list if filter(a)]
        remaining = [a for a in self.action_list if not filter(a)]
        self.action_list = remaining
        return extracted

    def copy(self):
        return PlayerAction(self.player, self.action_list.copy())

    def get_cost(self):
        return sum([action.get_cost() for action in self.action_list])

    def __repr__(self):
        return ','.join([str(a) for a in self.action_list])
    
    def serialize(self):
        return json.dumps([self.player] + [
            action.serialize() for action in self.action_list])

class ActionType(Enum):
    Upgrade = 1
    Defend = 2
    Move = 3
    Attack = 4
    Recruit = 5

    @classmethod
    def show(self, action_type):
        return {
            ActionType.Upgrade: 'UPG',
            ActionType.Defend: 'DEF',
            ActionType.Move: 'MOV',
            ActionType.Attack: 'ATK',
            ActionType.Recruit: 'REC'
        }[action_type]

class Position:
    def __init__(self, x, y):
        if not (0 <= x < board_size_x and 0 <= y < board_size_y):
            raise InvalidParameter("Position")
        self.x = x
        self.y = y

    @classmethod
    def from_literal(self, literal):
        try:
            x, y = literal
            x, y = map(int, [x, y])
        except:
            raise InvalidParameter("move literal")
        else:
            return Position(x - 1, y - 1)

    def get_delta(self, position):
        return PositionDelta(position.x - self.x, position.y - self.y)
    
    def __eq__(self, position):
        return self.x == position.x and self.y == position.y

    def get_new_position(self, position_delta):
        try:
            return Position(self.x + position_delta.dx, self.y + position_delta.dy)
        except InvalidParameter:
            return None

    def __str__(self):
        return f'{self.x + 1}{self.y + 1}'

    def __hash__(self):
        return self.x * 10 + self.y

    def serialize(self):
        return f'{self.x}{self.y}'

class Unit:
    display = ""
    level = 0
    trophy = 0
    
    def __init__(self, owner, skillset=None, flip_skillset=False):
        self.owner = owner
        self.perfect_skillset = potential_skillset_map[self.display].copy()
        if skillset is None:
            self.skillset = inborn_skillset_map[self.display].copy()
        else:
            self.skillset = skillset.copy()

        if flip_skillset:
            self.skillset.flip()

    def endow(self, skill):
        if self.perfect_skillset.has(skill):
            self.skillset.add(skill)
            return True
        return False
        
    def has_skill(self, skill):
        return self.skillset.has(skill)
    
    def has_potential_skill(self, skill):
        return self.potential_skillset().has(skill)

    def is_perfect(self):
        return self.skillset == self.perfect_skillset

    def is_promotion_ready(self):
        return not self.is_advanced() and self.is_perfect()

    def is_advanced(self):
        return type(self) not in promotion_map

    def potential_skillset(self):
        return self.ultimate_skillset().subtract(self.skillset)
    
    def ultimate_skillset(self):
        ultimate = self.perfect_skillset.copy()
        if self.is_promotion_ready():
            for creator in promotion_map[type(self)]:
                ultimate.union(potential_skillset_map[creator.display])
        return ultimate

    def get_promoted(self, skill):
        if not self.is_promotion_ready():
            return None

        creator = Unit.which_creator_has_skill(promotion_map[type(self)], skill)
        if creator is None:
            return None
        promoted = creator(self.owner, skillset=self.skillset)
        promoted.endow(skill)
        return promoted
    
    def duel(self, other):
        if self.level == other.level:
            return None
        elif self.level > other.level:
            return self
        else:
            return other

    @classmethod
    def which_creator_has_skill(self, creator_list, skill):
        for creator in creator_list:
            if potential_skillset_map[creator.display].has(skill):
                return creator

    @classmethod
    def create_from_skill(self, player, skill):
        if skill.delta.dx == skill.delta.dy == 0:
            creator = Wagon
        else:
            creator = self.which_creator_has_skill(
                [Rider, Soldier, Barbarian, Archer], skill)
        if creator is None:
            return None

        created = creator(player, SkillSet())
        created.endow(skill)
        return created

    def serialize(self):
        return json.dumps([type(self).__name__, self.owner, self.skillset.serialize()])

    @classmethod
    def deserialize(cls, payload):
        [typename, owner, p_skillset] = json.loads(payload)
        return eval(typename)(owner, SkillSet.deserialize(p_skillset))

class PositionDelta:
    def __init__(self, dx, dy):
        self.dx = dx
        self.dy = dy
    
    def __eq__(self, other):
        return self.dx == other.dx and self.dy == other.dy

class InvalidParameter(Exception):
    pass

class Skill:
    def __init__(self, position_delta):
        self.delta = position_delta
        if not (
                0 <= self.delta.dx + skillset_range < skillset_size and \
                0 <= self.delta.dy + skillset_range < skillset_size):
            raise InvalidParameter("Skill")
    
    def is_leap(self):
        return abs(self.delta.dx) > 1 or abs(self.delta.dy) > 1

class SkillSet:
    op_union = lambda a, b: a or b
    op_common = lambda a, b: a and b
    op_subtract = lambda a, b: a and not b

    def __init__(self, skill_list=[]):
        self.map = [[False] * skillset_size for i in range(skillset_size)]
        for skill in skill_list:
            self.add(skill)

    def apply(self, operator, skillset):
        self.map = [
            [
                operator(self.map[i][j], skillset.map[i][j]) 
                for j in range(skillset_size)
            ]
            for i in range(skillset_size)
        ]
        return self

    def copy(self):
        return SkillSet().union(self)

    def union(self, skillset):
        return self.apply(SkillSet.op_union, skillset)

    def subtract(self, skillset):
        return self.apply(SkillSet.op_subtract, skillset)

    def __eq__(self, other):
        return all(
            self.map[i][j] == other.map[i][j]
            for i in range(skillset_size)
            for j in range(skillset_size))

    def flip(self):
        self.map = [list(reversed(col)) for col in self.map]

    def add(self, skill):
        self.map[skill.delta.dx + skillset_range][skill.delta.dy + skillset_range] = True

    def has(self, skill):
        return self.map[skill.delta.dx + skillset_range][skill.delta.dy + skillset_range]

    def list_skills(self):
        return [
            Skill(PositionDelta(x - skillset_range, y - skillset_range))
            for x in range(skillset_size)
            for y in range(skillset_size)
            if self.map[x][y]
        ]
    
    def serialize(self):
        m = [[0] * skillset_size for i in range(skillset_size)]
        for i in range(skillset_size):
            for j in range(skillset_size):
                if self.map[i][j]:
                    m[i][j] = 1
        return json.dumps(m)

    @classmethod
    def deserialize(cls, payload):
        s = SkillSet()
        m = json.loads(payload)
        for i in range(skillset_size):
            for j in range(skillset_size):
                s.map[i][j] = m[i][j] == 1
        return s

class Rider(Unit):
    display = "RDR"
    level = 2
    trophy = 8

class Soldier(Unit):
    display = "SLD"
    level = 1
    trophy = 6

class Archer(Unit):
    display = "ACH"
    level = 1
    trophy = 5

class Barbarian(Unit):
    display = "BAR"
    level = 1
    trophy = 4

class Lancer(Unit):
    display = "LAN"
    level = 3
    trophy = 8

class Knight(Unit):
    display = "KNT"
    level = 3
    trophy = 8

class Swordsman(Unit):
    display = "SWD"
    level = 2
    trophy = 5

class Spearman(Unit):
    display = "SPR"
    level = 2
    trophy = 6

class Warrior(Unit):
    display = "WAR"
    level = 2
    trophy = 4

class King(Unit):
    display = "KING"
    level = 1
    trophy = 0

class Wagon(Unit):
    display = "WAG"
    level = 0
    trophy = 20

def convert_skill_list_map_to_skillset_map(skill_list_map):
    return {
        k: SkillSet([Skill(PositionDelta(x - skillset_range, y - skillset_range)) for x, y in v])
        for k, v in skill_list_map.items() 
    }

promotion_map = {
    Rider: [Lancer, Knight],
    Soldier: [Swordsman, Spearman],
    Archer: [Warrior, Spearman],
    Barbarian: [Warrior, Swordsman]
}

potential_skillset_map = convert_skill_list_map_to_skillset_map(skills.potential_skill_list_map)
inborn_skillset_map = convert_skill_list_map_to_skillset_map(skills.inborn_skill_list_map)

